enumerables.rb

```diff
@@ -0,0 +1,209 @@
+        return flag
+      end
+      my_each { |item| flag = false if (yield(item) == false) || yield(item).nil? }
+      return flag
+    end
+    flag
+  end
+
+  # my_any?
+  def my_any?(arg = nil)
+    flag = false
+    case arg
+    when Class
+      my_each { |item| flag = true if item.is_a? arg }
+    when Regexp
+      my_each { |item| flag = true if arg.match(item.to_s) }
+    when String
+      my_each { |item| flag = true if item.to_s.include? arg }
+    when Integer
+      my_each { |item| flag = true if item == arg }
+    end
+    if block_given? && arg.nil?
+      my_each { |item| flag = true if yield(item) }
+    elsif !block_given? && arg.nil?
+      arg = proc { |obj| obj }
+      my_each { |item| flag = true unless [nil, false].include?(item) }
+    end
+    flag
+  end
+
+  # my_none?
+  def my_none?(arg = nil)
+    flag = true
+    case arg
+    when Class
+      my_each { |item| flag = false if item.is_a? arg }
+    when Regexp
+      my_each { |item| flag = false if arg.match(item) }
+    when String
+      my_each { |item| flag = false if item.to_s.include? arg }
+    when Integer
+      my_each { |item| flag = false if item == arg }
+    end
+    if block_given? && arg.nil?
+      my_each { |item| flag = false if yield(item) }
+    elsif !block_given? && arg.nil?
+      arg = proc { |obj| obj }
+      my_each { |item| flag = false if arg.call(item) }
+    end
+    flag
+  end
+
+  # my_count
+  def my_count(arg = nil)
+    counter = 0
+    if arg
+      my_each { |item| counter += 1 if item == arg }
+    elsif block_given?
+      my_each { |item| counter += 1 if yield(item) }
+    else
+      my_each { counter += 1 }
+    end
+    counter
+  end
+
+  # my_map
+  def my_map(proc = nil, &block)
+    return to_enum(:my_map) unless block_given?
+
+    new_arr = []
+    if proc.is_a? Proc
+      my_each do |item|
+        new_item = proc.call(item)
+        new_arr << new_item
+      end
+      return new_arr
+    end
+
+    my_each do |item|
+      new_item = block.call(item)
+      new_arr << new_item
+    end
+    new_arr
+  end
+
+  # my_inject
+  def my_inject(init = nil, sign = nil)
+    raise LocalJumpError unless block_given? || sign || init || my_all?(String)
+
+    if init && sign
+      result = init
+      my_each { |item| result = result.send(sign, item) }
+    elsif init.is_a?(Symbol) || init.is_a?(String)
+      result = first
+      range_arr = to_a
+      (1..size - 1).my_each { |item| result = result.send(init.to_sym, range_arr[item]) }
+    elsif my_all?(String)
+      longest_word = 0
+      my_each { |item| longest_word = item.length if item.length > longest_word }
+      return longest_word
+    elsif block_given?
+      if init result = init
+        my_each { |item| result = yield(result, item) }
+      else
+        result = first
+        range_arr = to_a
+        (1..size - 1).my_each { |item| result = yield(result, range_arr[item]) }
+      end
+    end
+    result
+  end
+
+  # end Module
+end
```

Comment:
Your `my_inject` method returned the following errors:

```
Failures:

  1) enumerables #my_inject searches for the longest word in an array of strings
     Failure/Error: expect(words.my_inject(&search)).to eq words.inject(&search)
     
       expected: "blade"
            got: 5
     
       (compared using ==)
     # ./spec/enumerables_spec.rb:410:in `block (3 levels) in <top (required)>'

  2) enumerables #my_inject does not mutate the original array
     Failure/Error: array.my_inject { |num| num + 1 }
     
     NoMethodError:
       undefined method `init' for #<Array:0x00005637cd6c4d68>
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:414:in `block (3 levels) in <top (required)>'

  3) enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)
     
     NoMethodError:
       undefined method `init' for #<Array:0x00005637cd7bd670>
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:420:in `block (4 levels) in <top (required)>'

  4) enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
     Failure/Error: actual = range.my_inject { |prod, n| prod * n }
     
     NoMethodError:
       undefined method `init' for 5..50:Range
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:424:in `block (4 levels) in <top (required)>'

  5) enumerables #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
     
     NoMethodError:
       undefined method `init' for 5..50:Range
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:432:in `block (4 levels) in <top (required)>'
```

Here's an explanation of each of these error messages:

1)  Your `my_inject` is returning the wrong output here. It should return the value of the longest word, not the number of letters in it. You may test your code in the following manner to make sure it outputs the correct value:

```
search = proc { |memo, word| memo.length > word.length ? memo : word }
words = %w[dog door rod blade] 
words.my_inject(&search) # This should return "blade"
```

2) 3) 4) 5) Apparently, you have a bug in line 155 in your `enumerables.rb` file. Please make sure that the `init` method you are using here exists for the cases. You may use the following setup for testing these cases:

```
array = [5, 4, 3, 2, 1]
range =   Range.new(5, 50)
operation = proc { |sum, n| sum + n }
array_clone = array.clone 

# Make sure the following calls to `my_inject` don't produce an `method init not found` error:
array.my_inject { |num| num + 1 }
array.my_inject(&operation)
range.my_inject { |prod, n| prod * n }
range.my_inject(4) { |prod, n| prod * n }
```